File: Design Data Structures/MedianDataStreamLC295.java
import java.util.*;

public class MedianDataStreamLC295 {
    Queue<Integer> large;
    Queue<Integer> small;

    /** initialize your data structure here. */
    public MedianDataStreamLC295() {
        large = new PriorityQueue<Integer>();
        small = new PriorityQueue<Integer>();
    }
    
    public void addNum(int num) {
        large.add(num);
        small.add(-large.poll());
        if (small.size() > large.size()) {
            large.add(-small.poll());
        }
    }
    
    public double findMedian() {
        return large.size() > small.size()
               ? large.peek()
               : (large.peek() - small.peek()) / 2.0;
    }
}

/**
 * Your MedianFinder object will be instantiated and called as such:
 * MedianFinder obj = new MedianFinder();
 * obj.addNum(num);
 * double param_2 = obj.findMedian();
 */

File: Design Data Structures/MedianDataStreamLC295.py
from heapq import heappush, heappushpop, heappop

class MedianFinder:

    def __init__(self):
        """
        initialize your data structure here.
        """
        self.minHeap = []
        self.maxHeap = []

    def addNum(self, num: int) -> None:
        small, large = self.minHeap, self.maxHeap
        
        heappush(small, -heappushpop(large, -num))
        if len(small) > len(large):
            heappush(large, -heappop(small))        

    def findMedian(self) -> float:
        small, large = self.minHeap, self.maxHeap
        if (len(small) + len(large))%2 == 0:
            return (-float(large[0]) + float(small[0]))/2
        return -float(large[0])
            
 

# Your MedianFinder object will be instantiated and called as such:
# obj = MedianFinder()
# obj.addNum(num)
# param_2 = obj.findMedian()